README.md
# to-do-list-clojure

A Clojure To-Do List Application. Can add dates and entries to those dates.

## Usage

core.clj is the file that runs entire application. Run (main) to start application.

## License

Copyright © 2021 Matthew Lad

This program and the accompanying materials are made available under the
terms of the Eclipse Public License 2.0 which is available at
http://www.eclipse.org/legal/epl-2.0.

This Source Code may also be made available under the following Secondary
Licenses when the conditions for such availability set forth in the Eclipse
Public License, v. 2.0 are satisfied: GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or (at your
option) any later version, with the GNU Classpath Exception which is available
at https://www.gnu.org/software/classpath/license.html.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2021-04-29
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2021-04-29
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/to-do-list-clojure/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/to-do-list-clojure/compare/0.1.0...0.1.1


test/to_do_list_clojure/core_test.clj
(ns to-do-list-clojure.core-test
  (:require [clojure.test :refer :all]
            [to-do-list-clojure.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


src/to_do_list_clojure/data.clj
;;;;----------------------------------------------------------;;;;
;;;;     Matthew Lad                                          ;;;;
;;;;     To-Do List Application                               ;;;;
;;;;     4/29/2021                                            ;;;;
;;;;----------------------------------------------------------;;;;

(ns to-do-list-clojure.data
  (:require [clojure.string :as str])
  (:import (java.util Date)
           (java.text SimpleDateFormat)))


;; The main database for storing to-do list data.
;; It is a map where the date is keyed to vectors containing
;; entries for those dates.
(def main-data {})


(defn wipe-main-data
  "Erases all data in main-list.
   Returns main-list"
  []
  (def main-data {})
  main-data)


(defn print-main-list
  "Prints out the contents of main-list in a very simple format.
   Used for testing/debugging more than anything"
  []
  (for [date main-data]
    (println date)))


(defn millis-to-date
  "Converts a date in milliseconds to a string of
   format 'mm/dd/yyyy'"
  [date-millis]
  (let [date-obj (new Date date-millis)
        sdf (SimpleDateFormat. "MM/dd/yyyy")
        date-str (.format sdf date-obj)]
    (.toString date-str)))


(defn date-to-millis
  "Converts a date string of format 'mm/dd/yyyy'
   into it's equivalent milliseconds since the epoch."
  [date-str]
  (let [sdf (SimpleDateFormat. "MM/dd/yyyy")
        date (.parse sdf date-str)
        millis (.getTime date)]
    millis))


(defn format-date
  "The function to create a Date out of raw strings.
   Takes in a date in the format of 'mm/dd/yyyy' and returns
   a map where the key is the date in milliseconds since the
   epoch and the value is an empty vector. Example:
   (create-date '02/09/2001')
   => {981698400000 [], etc.}"
  [date-string]
  (assoc {} (date-to-millis date-string) []))


(defn format-entry
  "The function to create an Entry.
   Takes in an importance number and entry subject string and
   returns a vector containing those input items.
   Example: (create-entry 1 'Hi') => [1 'Hi']"
  [importance-num entry-string]
  [importance-num entry-string])


(defn add-date
  "This function adds a date to the main-data map.
   It takes in a date as a map {millis []}
   If the date already exists, does nothing.
   If the date doesn't exist, add it to main-data"
  [date-map]
  (if (contains? main-data (first (keys date-map)))
    nil
    (def main-data (merge main-data date-map))))


(defn add-entry
  "1. have it add the date to the main-list
   2. have it add the entry to the main-list"
  [date-str imp-num entry-str]
  (let [date-millis (first (keys (format-date date-str)))
        frmtd-date (format-date date-str)
        frmtd-entry (format-entry imp-num entry-str)]
    (add-date (format-date date-str))  ; This adds the date if it doesn't exist already.
    (def main-data ;; This adds the entry to the correct date
      (assoc-in main-data [date-millis]
                (conj (get main-data date-millis) frmtd-entry)))))


(defn get-entries
  "Returns all the entries for a particular date.
   Returns as [[importance 'subject'] [importance 'subject'] ...]"
  [date-str]
  (let [date-millis (first (keys (format-date date-str)))]
    (get main-data date-millis)))


(defn get-dates-string
  "Returns all the dates in main-data in a vector of strings."
  []
  (vec (map millis-to-date (keys main-data))))




src/to_do_list_clojure/MainGUI.java
package to_do_list_clojure;

public class MainGUI extends javax.swing.JFrame {

    public MainGUI() {
        initComponents();
    }


    private void initComponents() {

        dateScrollPane = new javax.swing.JScrollPane();
        dateList = new javax.swing.JList<>();
        entryListLabel = new javax.swing.JLabel();
        dateListLabel = new javax.swing.JLabel();
        selectDateButton = new javax.swing.JButton();
        tipsScrollPane = new javax.swing.JScrollPane();
        tipsTextPane = new javax.swing.JTextPane();
        entriesScrollPane = new javax.swing.JScrollPane();
        entriesList = new javax.swing.JList<>();
        completedButton = new javax.swing.JButton();
        toDoListTitleLabel = new javax.swing.JLabel();

        setDefaultCloseOperation(javax.swing.WindowConstants.EXIT_ON_CLOSE);
        setTitle("To-Do List");

        dateList.setModel(new javax.swing.AbstractListModel<String>() {
            String[] strings = { "Item 1", "Item 2", "Item 3", "Item 4", "Item 5" };
            public int getSize() { return strings.length; }
            public String getElementAt(int i) { return strings[i]; }
        });
        dateScrollPane.setViewportView(dateList);

        entryListLabel.setFont(new java.awt.Font("Segoe UI", 1, 18)); // NOI18N
        entryListLabel.setText("Entries");

        dateListLabel.setFont(new java.awt.Font("Segoe UI", 1, 18)); // NOI18N
        dateListLabel.setText("Date");

        selectDateButton.setFont(new java.awt.Font("Segoe UI", 0, 12)); // NOI18N
        selectDateButton.setText("Select");
        selectDateButton.addActionListener(new java.awt.event.ActionListener() {
            public void actionPerformed(java.awt.event.ActionEvent evt) {
                selectDateButtonActionPerformed(evt);
            }
        });

        tipsTextPane.setEditable(false);
        tipsTextPane.setBackground(new java.awt.Color(240, 240, 240));
        tipsTextPane.setBorder(javax.swing.BorderFactory.createBevelBorder(javax.swing.border.BevelBorder.RAISED));
        tipsTextPane.setFont(new java.awt.Font("Segoe UI", 0, 12)); // NOI18N
        tipsTextPane.setText("TIPS:");
        tipsScrollPane.setViewportView(tipsTextPane);

        entriesList.setModel(new javax.swing.AbstractListModel<String>() {
            String[] strings = { "Item 1", "Item 2", "Item 3", "Item 4", "Item 5" };
            public int getSize() { return strings.length; }
            public String getElementAt(int i) { return strings[i]; }
        });
        entriesScrollPane.setViewportView(entriesList);

        completedButton.setFont(new java.awt.Font("Segoe UI", 0, 12)); // NOI18N
        completedButton.setText("Mark Complete");
        completedButton.addActionListener(new java.awt.event.ActionListener() {
            public void actionPerformed(java.awt.event.ActionEvent evt) {
                completedButtonActionPerformed(evt);
            }
        });

        toDoListTitleLabel.setFont(new java.awt.Font("Segoe UI", 1, 24)); // NOI18N
        toDoListTitleLabel.setText("To-Do List");

        javax.swing.GroupLayout layout = new javax.swing.GroupLayout(getContentPane());
        getContentPane().setLayout(layout);
        layout.setHorizontalGroup(
                layout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING)
                        .addGroup(javax.swing.GroupLayout.Alignment.TRAILING, layout.createSequentialGroup()
                                .addGap(20, 20, 20)
                                .addGroup(layout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING)
                                        .addComponent(tipsScrollPane, javax.swing.GroupLayout.PREFERRED_SIZE, 147, javax.swing.GroupLayout.PREFERRED_SIZE)
                                        .addComponent(toDoListTitleLabel, javax.swing.GroupLayout.PREFERRED_SIZE, 123, javax.swing.GroupLayout.PREFERRED_SIZE))
                                .addGap(18, 18, Short.MAX_VALUE)
                                .addGroup(layout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING)
                                        .addGroup(layout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING)
                                                .addComponent(dateScrollPane, javax.swing.GroupLayout.PREFERRED_SIZE, 95, javax.swing.GroupLayout.PREFERRED_SIZE)
                                                .addGroup(javax.swing.GroupLayout.Alignment.TRAILING, layout.createSequentialGroup()
                                                        .addComponent(dateListLabel, javax.swing.GroupLayout.PREFERRED_SIZE, 51, javax.swing.GroupLayout.PREFERRED_SIZE)
                                                        .addGap(23, 23, 23)))
                                        .addGroup(layout.createSequentialGroup()
                                                .addGap(10, 10, 10)
                                                .addComponent(selectDateButton, javax.swing.GroupLayout.PREFERRED_SIZE, 69, javax.swing.GroupLayout.PREFERRED_SIZE)
                                                .addGap(26, 26, 26)))
                                .addGroup(layout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING)
                                        .addGroup(layout.createSequentialGroup()
                                                .addGap(156, 156, 156)
                                                .addComponent(entryListLabel, javax.swing.GroupLayout.PREFERRED_SIZE, 66, javax.swing.GroupLayout.PREFERRED_SIZE))
                                        .addGroup(layout.createSequentialGroup()
                                                .addGap(9, 9, 9)
                                                .addComponent(entriesScrollPane, javax.swing.GroupLayout.PREFERRED_SIZE, 546, javax.swing.GroupLayout.PREFERRED_SIZE))
                                        .addGroup(layout.createSequentialGroup()
                                                .addGap(19, 19, 19)
                                                .addComponent(completedButton, javax.swing.GroupLayout.PREFERRED_SIZE, 126, javax.swing.GroupLayout.PREFERRED_SIZE)))
                                .addGap(31, 31, 31))
        );
        layout.setVerticalGroup(
                layout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING)
                        .addGroup(javax.swing.GroupLayout.Alignment.TRAILING, layout.createSequentialGroup()
                                .addContainerGap()
                                .addGroup(layout.createParallelGroup(javax.swing.GroupLayout.Alignment.BASELINE)
                                        .addComponent(dateListLabel)
                                        .addComponent(entryListLabel))
                                .addPreferredGap(javax.swing.LayoutStyle.ComponentPlacement.RELATED)
                                .addGroup(layout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING)
                                        .addGroup(layout.createSequentialGroup()
                                                .addComponent(toDoListTitleLabel, javax.swing.GroupLayout.PREFERRED_SIZE, 55, javax.swing.GroupLayout.PREFERRED_SIZE)
                                                .addGap(65, 65, 65)
                                                .addComponent(tipsScrollPane, javax.swing.GroupLayout.PREFERRED_SIZE, 101, javax.swing.GroupLayout.PREFERRED_SIZE)
                                                .addGap(0, 97, Short.MAX_VALUE))
                                        .addComponent(entriesScrollPane)
                                        .addComponent(dateScrollPane, javax.swing.GroupLayout.Alignment.TRAILING))
                                .addPreferredGap(javax.swing.LayoutStyle.ComponentPlacement.UNRELATED)
                                .addGroup(layout.createParallelGroup(javax.swing.GroupLayout.Alignment.BASELINE)
                                        .addComponent(completedButton)
                                        .addComponent(selectDateButton))
                                .addContainerGap())
        );

        pack();
    }// </editor-fold>

    private void selectDateButtonActionPerformed(java.awt.event.ActionEvent evt) {
        // TODO add your handling code here:
    }

    private void completedButtonActionPerformed(java.awt.event.ActionEvent evt) {
        // TODO add your handling code here:
    }

    public static void main(String args[]) {
        java.awt.EventQueue.invokeLater(new Runnable() {
            public void run() {
                new MainGUI().setVisible(true);
            }
        });
    }

    // Variables declaration
    private javax.swing.JButton completedButton;
    private javax.swing.JList<String> dateList;
    private javax.swing.JLabel dateListLabel;
    private javax.swing.JScrollPane dateScrollPane;
    private javax.swing.JList<String> entriesList;
    private javax.swing.JScrollPane entriesScrollPane;
    private javax.swing.JLabel entryListLabel;
    private javax.swing.JButton selectDateButton;
    private javax.swing.JScrollPane tipsScrollPane;
    private javax.swing.JTextPane tipsTextPane;
    private javax.swing.JLabel toDoListTitleLabel;
    // End of variables declaration
}



src/to_do_list_clojure/gui.clj
;;;;----------------------------------------------------------;;;;
;;;;     Matthew Lad                                          ;;;;
;;;;     To-Do List Application                               ;;;;
;;;;     4/29/2021                                            ;;;;
;;;;                                                          ;;;;
;;;;     This file creates the GUI.                           ;;;;
;;;;----------------------------------------------------------;;;;

(ns to-do-list-clojure.gui
  (:require [to-do-list-clojure.data :as db])
  (:import (javax.swing JTextPane            JScrollPane
                        JLabel               JList
                        JButton              JFrame
                        BorderFactory        JTextField
                        ListSelectionModel   DefaultListModel)

           (java.awt Font                   Color
                     GridBagConstraints     GridBagLayout
                     Insets)

           (javax.swing.border BevelBorder)
           (java.awt.event ActionListener)))

;;;----------------------------------------------------------;;;
;;; Creates a primary-frame for the to-do list application   ;;;
;;;----------------------------------------------------------;;;
(def primary-frame
  (doto (new JFrame)
    (.setTitle "Lad's To-Do-List!")
    (.setDefaultCloseOperation JFrame/EXIT_ON_CLOSE)
    (.setLayout (new GridBagLayout))))

;;;----------------------------------------------------------;;;
;;; Creates some initial global variables (components)       ;;;
;;;----------------------------------------------------------;;;
(def date-input-text-field (new JTextField))
(def entry-input-text-field (new JTextField))
(def date-list-model (new DefaultListModel))
(def date-list (new JList date-list-model))
(def entry-list-model (new DefaultListModel))
(def entries-list (new JList entry-list-model))


;;;----------------------------------------------------------;;;
;;; Action Functions for each of the Buttons                 ;;;
;;;----------------------------------------------------------;;;

;;; Action to add an entry to the to-do-list.
;;; This action is connected to the add entry button.
(defn add-entry-action []
  (let [date-input (str (.getText date-input-text-field))
        entry-input (str (.getText entry-input-text-field))]
    (db/add-entry date-input 1 entry-input)
    (.addElement date-list-model date-input)

    ;; entry-input isn't working right.
    ;; Adding multiples of the same date is possible
    ;; but shouldn't be.
    
    ))

;;;
(defn delete-entry-action []
  )

;;;
(defn select-date-action []
  )

;;;
(defn mark-complete-action []
  )


;;;----------------------------------------;;;
;;; Function to Initializes all components ;;;
;;;----------------------------------------;;;
(defn initialize-primary-frame
  "Initializes all the components and adds them to the frame."
  [frame]

  ;;;-------------------------------------------------;;;
  ;;; Date List Data, Label, Constraints, Scroll-Pane ;;;
  ;;;-------------------------------------------------;;;

  (def date-list-model
    (doto (new DefaultListModel)
      (.addElement "mm/dd/yyyy")))

  ;; Creates a list component for the dates
  (def date-list
    (doto (new JList date-list-model)
      (.setSelectionMode ListSelectionModel/SINGLE_INTERVAL_SELECTION)
      (.setLayoutOrientation JList/VERTICAL)))

  ;; Creates a scroll pane component for the dates
  (def date-scroll-pane
    (doto (new JScrollPane)
      (.setViewportView date-list)))

  ;; Creates the Grid Bag Constraints for the date scroll component
  (def date-scroll-pane-grid-bag
    (let [grid-bag (new GridBagConstraints)]
      (set! (. grid-bag -gridx) 2)
      (set! (. grid-bag -gridy) 1)
      (set! (. grid-bag -gridwidth) 2)
      (set! (. grid-bag -gridheight) 2)
      (set! (. grid-bag -fill) GridBagConstraints/BOTH)
      (set! (. grid-bag -ipadx) 56)
      (set! (. grid-bag -ipady) 262)
      (set! (. grid-bag -anchor) GridBagConstraints/NORTHWEST)
      (set! (. grid-bag -weightx) 1.0)
      (set! (. grid-bag -weighty) 1.0)
      (set! (. grid-bag -insets) (new Insets 6 18 0 0))
      grid-bag))

  ;; Adds the dates scroll pane and the grid bag constraints to the frame
  (doto frame
    (.add date-scroll-pane date-scroll-pane-grid-bag))

  ;; Creates a label for the date list
  (def date-list-label
    (doto (new JLabel)
      (.setFont (new Font "Segoe UI" 1 18))
      (.setText "Date")))

  ;; Creates grid-bag constraints for the date list label
  (def date-list-label-grid-bag
    (let [grid-bag (new GridBagConstraints)]
      (set! (. grid-bag -gridx) 2)
      (set! (. grid-bag -gridy) 0)
      (set! (. grid-bag -ipadx) 11)
      (set! (. grid-bag -anchor) GridBagConstraints/NORTHWEST)
      (set! (. grid-bag -insets) (new Insets 10 39 0 0))
      grid-bag))

  ;; Adds the date list label and date list label constraints to the frame
  (doto frame
    (.add date-list-label date-list-label-grid-bag))


  ;;;--------------------------------------------------;;;
  ;;; Entry List Data, Label, Constraints, Scroll-Pane ;;;
  ;;;--------------------------------------------------;;;

  ;; Creates a label component for the entries list
  (def entry-list-label
    (doto (new JLabel)
      (.setFont (new Font "Segoe UI" 1 18))
      (.setText "Entries")))

  ;; Creates the grid bag constraints for the entry list label
  (def entry-list-label-grid-bag
    (let [grid-bag (new GridBagConstraints)]
      (set! (. grid-bag -gridx) 5)
      (set! (. grid-bag -gridy) 0)
      (set! (. grid-bag -ipadx) 8)
      (set! (. grid-bag -anchor) GridBagConstraints/NORTHWEST)
      (set! (. grid-bag -insets) (new Insets 10 22 0 0))
      grid-bag))

  ;; Adds the entry list label and it's constraints
  (doto frame
    (.add entry-list-label entry-list-label-grid-bag))

  ;; Creates data to go into the entries list
  (def entries-list-model
    (doto (new DefaultListModel)
      (.addElement "Thing that needs to get done.")))

  ;; Creates a list component for the entries list
  (def entries-list
    (doto (new JList entries-list-model)))

  ;; Creates a scroll pane for the entries
  (def entries-scroll-pane
    (doto (new JScrollPane)
      (.setViewportView entries-list)))

  ;; Creates constraints for the entries scroll pane
  (def entries-scroll-pane-grid-bag
    (let [grid-bag (new GridBagConstraints)]
      (set! (. grid-bag -gridx) 4)
      (set! (. grid-bag -gridy) 1)
      (set! (. grid-bag -gridwidth) 5)
      (set! (. grid-bag -gridheight) 2)
      (set! (. grid-bag -fill) GridBagConstraints/BOTH)
      (set! (. grid-bag -ipadx) 409)
      (set! (. grid-bag -ipady) 262)
      (set! (. grid-bag -anchor) GridBagConstraints/NORTHWEST)
      (set! (. grid-bag -weightx) 1.0)
      (set! (. grid-bag -weighty) 1.0)
      (set! (. grid-bag -insets) (new Insets 6 19 0 16))
      grid-bag))

  ;; Adds the entries scroll pane and scroll pane constraints to the frame.
  (doto frame
    (.add entries-scroll-pane entries-scroll-pane-grid-bag))


  ;;;--------------------------------------------------;;;
  ;;; Select date, completed, add, and delete buttons  ;;;
  ;;;--------------------------------------------------;;;

  ;; Creates a button for selecting the date
  (def select-date-btn
    (doto (new JButton)
      (.setFont (new Font "Segoe UI" 0 12))
      (.setText "Select")))

  ;; Need to add the Button Action Listener!!

  ;; Creates constraints for the select button
  (def select-date-grid-bag
    (let [grid-bag (new GridBagConstraints)]
      (set! (. grid-bag -gridx) 2)
      (set! (. grid-bag -gridy) 5)
      (set! (. grid-bag -gridwidth) 2)
      (set! (. grid-bag -ipadx) 4)
      (set! (. grid-bag -anchor) GridBagConstraints/NORTHWEST)
      (set! (. grid-bag -insets) (new Insets 15 28 10 0))
      grid-bag))

  ;; Adds the select date button and it's constraints to the frame.
  (doto frame
    (.add select-date-btn select-date-grid-bag))

  ;; Creates a button for indicating completed entries
  (def completed-btn
    (doto (new JButton)
      (.setFont (new Font "Segoe UI" 0 12))
      (.setText "Mark Complete")))

  ;; Need to add Button Action Listener!

  ;; Completed Button Constraints
  (def completed-btn-grid-bag
    (let [grid-bag (new GridBagConstraints)]
      (set! (. grid-bag -gridx) 4)
      (set! (. grid-bag -gridy) 5)
      (set! (. grid-bag -ipadx) 11)
      (set! (. grid-bag -anchor) GridBagConstraints/NORTHWEST)
      (set! (. grid-bag -insets) (new Insets 15 18 10 0))
      grid-bag))

  ;; Adds the completed button and it's constraints to the frame.
  (doto frame
    (.add completed-btn completed-btn-grid-bag))

  ;; Creates a button for adding entries
  (def add-entry-btn
    (doto (new JButton)
      (.setFont (new Font "Segoe UI" 0 12))
      (.setText "Add Entry")))

  ;; Add Entry Button Constraints
  (def add-entry-btn-grid-bag
    (let [grid-bag (new GridBagConstraints)]
      (set! (. grid-bag -gridx) 5)
      (set! (. grid-bag -gridy) 5)
      (set! (. grid-bag -gridwidth) 2)
      (set! (. grid-bag -anchor) GridBagConstraints/NORTHWEST)
      (set! (. grid-bag -insets) (new Insets 15 18 10 0))
      grid-bag))

  ;; Adds the add-entry button and it's constraints to the frame.
  (doto frame
    (.add add-entry-btn add-entry-btn-grid-bag))

  ;; Creates a button for deleting entries
  (def delete-entry-btn
    (doto (new JButton)
      (.setFont (new Font "Segoe UI" 0 12))
      (.setText "Delete Entry")))

  ;; Need to add Button Action Listener!

  ;; delete entries Button Constraints
  (def delete-entry-btn-grid-bag
    (let [grid-bag (new GridBagConstraints)]
      (set! (. grid-bag -gridx) 7)
      (set! (. grid-bag -gridy) 5)
      (set! (. grid-bag -anchor) GridBagConstraints/NORTHWEST)
      (set! (. grid-bag -insets) (new Insets 15 18 10 0))
      grid-bag))

  ;; Adds the delete-entry button and it's constraints to the frame.
  (doto frame
    (.add delete-entry-btn delete-entry-btn-grid-bag))


  ;;;--------------------------------------------------;;;
  ;;; Tips Text-Pane, scroll-pane, and constraints     ;;;
  ;;;--------------------------------------------------;;;

  ;; Creates a tips text-pane component
  (def tips-text-pane
    (doto (new JTextPane)
      (.setEditable false)
      (.setBackground (new Color 240 240 240))
      (.setFont (new Font "Segoe UI" 0 12))
      (.setText "TIPS:")
      (.setBorder (BorderFactory/createBevelBorder BevelBorder/RAISED))))

  ;; Creates a scroll pane for the tips text pane
  (def tips-scroll-pane
    (doto (new JScrollPane)
      (.setViewportView tips-text-pane)))

  ;; Creates constraints for the tips text pane
  (def tips-text-grid-bag
    (let [grid-bag (new GridBagConstraints)]
      (set! (. grid-bag -gridx) 0)
      (set! (. grid-bag -gridy) 2)
      (set! (. grid-bag -gridwidth) 2)
      (set! (. grid-bag -gridheight) 3)
      (set! (. grid-bag -fill) GridBagConstraints/BOTH)
      (set! (. grid-bag -ipadx) 102)
      (set! (. grid-bag -ipady) 181)
      (set! (. grid-bag -anchor) GridBagConstraints/NORTHWEST)
      (set! (. grid-bag -weightx) 1.0)
      (set! (. grid-bag -weighty) 1.0)
      (set! (. grid-bag -insets) (new Insets 65 17 0 0))
      grid-bag))

  ;; Adds the tips scroll pane and it's constraints to the frame.
  (doto frame
    (.add tips-scroll-pane tips-text-grid-bag))


  ;;;--------------------------------------------------;;;
  ;;; Title label and it's constraints.                ;;;
  ;;;--------------------------------------------------;;;

  ;; Creates a label component for the title
  (def title-label
    (doto (new JLabel)
      (.setFont (new Font "Segoe UI" 1 24))
      (.setText "To-Do List")))

  ;; Creates constraints for the title Label
  (def title-label-grid-bag
    (let [grid-bag (new GridBagConstraints)]
      (set! (. grid-bag -gridx) 0)
      (set! (. grid-bag -gridy) 1)
      (set! (. grid-bag -ipadx) 7)
      (set! (. grid-bag -ipady) 23)
      (set! (. grid-bag -anchor) GridBagConstraints/NORTHWEST)
      (set! (. grid-bag -insets) (new Insets 6 17 0 0))
      grid-bag))

  ;; Adds the title label and it's constraints to the frame
  (doto frame
    (.add title-label title-label-grid-bag))


  ;;;---------------------------------------------------------;;;
  ;;; Date input and entry input text fields and constraints  ;;;
  ;;;---------------------------------------------------------;;;

  ;; Creates a date input text field
  (def date-input-text-field
    (doto (new JTextField)
      (.setText "mm/dd/yyyy")))

  ;; Creates constraints for the date input text field
  (def date-input-grid-bag
    (let [grid-bag (new GridBagConstraints)]
      (set! (. grid-bag -gridx) 2)
      (set! (. grid-bag -gridy) 4)
      (set! (. grid-bag -gridwidth) 2)
      (set! (. grid-bag -ipadx) 72)
      (set! (. grid-bag -ipady) 10)
      (set! (. grid-bag -anchor) GridBagConstraints/NORTHWEST)
      (set! (. grid-bag -insets) (new Insets 10 18 0 0))
      grid-bag))

  ;; Adds the date input text field and it's constraints to the frame
  (doto frame
    (.add date-input-text-field date-input-grid-bag))

  ;; Creates an entry input text field
  (def entry-input-text-field
    (doto (new JTextField)
      (.setText "Fold the laundry.")))

  ;; Creates constraints for the entry input text field
  (def entry-input-grid-bag
    (let [grid-bag (new GridBagConstraints)]
      (set! (. grid-bag -gridx) 4)
      (set! (. grid-bag -gridy) 4)
      (set! (. grid-bag -gridwidth) 5)
      (set! (. grid-bag -ipadx) 426)
      (set! (. grid-bag -ipady) 10)
      (set! (. grid-bag -anchor) GridBagConstraints/NORTHWEST)
      (set! (. grid-bag -insets) (new Insets 10 18 0 16))
      grid-bag))

  ;; Adds the entry input text field and it's constraints to the frame
  (doto frame
    (.add entry-input-text-field entry-input-grid-bag))


  ;;;----------------------------------------------------------;;;
  ;;; Creates action listeners for each of the buttons         ;;;
  ;;; These listeners call action-functions.                   ;;;
  ;;;----------------------------------------------------------;;;
  (def add-entry-action-listener
    (proxy [ActionListener] []
      (actionPerformed [event] (add-entry-action))))

  (def delete-entry-action-listener
    (proxy [ActionListener] []
      (actionPerformed [event] (delete-entry-action))))

  (def select-date-action-listener
    (proxy [ActionListener] []
      (actionPerformed [event] (select-date-action))))

  (def mark-complete-action-listener
    (proxy [ActionListener] []
      (actionPerformed [event] (mark-complete-action))))


  ;;;----------------------------------------------------------;;;
  ;;; Adds action listeners to each of the buttons.            ;;;
  ;;;----------------------------------------------------------;;;
  (.addActionListener add-entry-btn add-entry-action-listener)
  (.addActionListener delete-entry-btn delete-entry-action-listener)
  (.addActionListener select-date-btn select-date-action-listener)
  (.addActionListener completed-btn mark-complete-action-listener)

  ) ;; End of initialize function.


;;;----------------------------------------------------------------------;;;
;;; Function to run the main GUI. It initializes and displays the gui.   ;;;
;;;----------------------------------------------------------------------;;;
(defn run []
  (initialize-primary-frame primary-frame)
  (doto primary-frame (.pack) (.setVisible true)))




